tests/factories.py

```diff
@@ -50,6 +51,62 @@ def ensure_workspace(workspace_or_name):
     raise Exception(f"Invalid workspace: {workspace_or_name})")  # pragma: nocover
 
 
+# get_output_metadata is imported from job-runner
+def get_output_metadata(
+    abspath, level, job_id, job_request, action, commit, excluded, message=None
+):
+    stat = abspath.stat()
+    with abspath.open("rb") as fp:
+        content_hash = file_digest(fp, "sha256").hexdigest()
+
+    return {
+        "level": level,
+        "job_id": job_id,
+        "job_request": job_request,
+        "action": action,
+        "commit": commit,
+        "size": stat.st_size,
+        "timestamp": stat.st_mtime,
+        "content_hash": content_hash,
+        "excluded": excluded,
+        "message": message,
+    }
+
+
+def update_manifest(workspace, files=None):
+    """Write a manfiest based on the files currently in the directory.
+
+    Make up action, job ids and commits.
+    """
+    workspace = ensure_workspace(workspace)
+    root = workspace.root()
+    manifest_path = root / "metadata/manifest.json"
+
+    if manifest_path.exists():
+        manifest = json.loads(manifest_path.read_text())
+        manifest["workspace"] = workspace.name
+        manifest.setdefault("outputs", {})
+    else:
+        manifest = {"workspace": workspace.name, "repo": None, "outputs": {}}
+
+    if files is None:  # pragma: nocover
+        files = [f.relative_to(root) for f in root.glob("**/*") if f.is_file()]
```

Comment:
Should this skip the metadata and logs dirs? I guess it doesn't really matter for tests and local workspaces

Reply:
oh, yes we should